Hook up general convolution to convolution_backward

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69584
* #69044
* #68899

Differential Revision: [D32936380](https://our.internmc.facebook.com/intern/diff/D32936380)

jbschlosser committed:
Hook up general convolution to convolution_backward

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32936380).

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1894,6 +1894,15 @@
   indices: non_differentiable
 
+- name: convolution(Tensor input, Tensor weight, Tensor? bias, int[] stride, int[] padding, int[] dilation, bool transposed, int[] output_padding, int groups) -> Tensor
+  input, weight, bias: "grad.defined() ? convolution_backward(grad, input, weight, bias->sizes(), stride, padding, dilation, transposed, output_padding, groups, grad_input_mask) : std::tuple<Tensor, Tensor, Tensor>()"
```
nit: Maybe it would be nicer to move the grad.defined() check inside convolution_backward ?
Or there is a reason why this is not possible?

jbschlosser replied:
I don't think it's possible - trying to do this gives me an error in the generated autograd code before the call to `convolution_backward` happens.

albanD replied:
Ok, non-CompositeImplicit function cannot take undefined Tensors as input.  LGTM then

jbschlosser committed:
Update on "Hook up general convolution to convolution_backward"


Differential Revision: [D32936380](https://our.internmc.facebook.com/intern/diff/D32936380)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32936380).

albanD reviewed (approved):
SGTM

jbschlosser committed:
Update on "Hook up general convolution to convolution_backward"


Differential Revision: [D32936380](https://our.internmc.facebook.com/intern/diff/D32936380)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32936380).

jbschlosser committed:
Update on "Hook up general convolution to convolution_backward"


Differential Revision: [D32936380](https://our.internmc.facebook.com/intern/diff/D32936380)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32936380).

albanD commented:
cc @soulitzer FYI this is the PR removing the number of conv we need to handle to 1

jbschlosser committed:
Update on "Hook up general convolution to convolution_backward"


Differential Revision: [D32936380](https://our.internmc.facebook.com/intern/diff/D32936380)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32936380).

jbschlosser committed:
Update on "Hook up general convolution to convolution_backward"


Differential Revision: [D32936380](https://our.internmc.facebook.com/intern/diff/D32936380)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32936380).

jbschlosser committed:
Update on "Hook up general convolution to convolution_backward"


Differential Revision: [D32936380](https://our.internmc.facebook.com/intern/diff/D32936380)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32936380).
